File: bindings/swift/Sources/Proven/SafeLRU.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// LRU cache delegated to libproven FFI.
///
/// Cache management is performed by the formally verified Idris 2 core
/// via the Zig FFI bridge. This class manages the lifecycle of the
/// C-allocated LRU cache.

import CProven

/// LRU cache backed by libproven. Keys are UInt64, values are Int64.
public final class SafeLRU {
    private let cache: UnsafeMutablePointer<ProvenLRUCache>

    /// Create an LRU cache with the given capacity.
    /// Returns nil if allocation fails.
    public init?(capacity: Int) {
        guard let ptr = proven_lru_create(capacity) else {
            return nil
        }
        self.cache = ptr
    }

    deinit {
        proven_lru_free(cache)
    }

    /// Get a value from the cache by key.
    /// Returns .failure if the key is not present or an error occurs.
    public func get(_ key: UInt64) -> Result<Int64, ProvenError> {
        let result = proven_lru_get(cache, key)
        if let error = ProvenError.fromStatus(result.status) {
            return .failure(error)
        }
        return .success(result.value)
    }

    /// Put a key-value pair into the cache.
    /// Evicts the least recently used entry if at capacity.
    @discardableResult
    public func put(_ key: UInt64, _ value: Int64) -> Result<Void, ProvenError> {
        let status = proven_lru_put(cache, key, value)
        if let error = ProvenError.fromStatus(status) {
            return .failure(error)
        }
        return .success(())
    }
}

File: bindings/nim/src/proven/safe_lru.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe LRU (Least Recently Used) cache.
# Thin wrapper over libproven FFI -- all logic lives in Idris.
#
# Note: The C API uses uint64 keys and int64 values.  This wrapper
# exposes the same key/value types.  For arbitrary key types, callers
# should hash them to uint64 externally.

import std/options
import lib_proven

type
  LruCache* = object
    ## LRU cache with bounded capacity, backed by libproven's verified
    ## implementation.  Keys are uint64 and values are int64.
    handle: ptr ProvenLRUCache

  LruCacheError* = object of CatchableError
    ## Error raised when LRU cache creation fails.

proc `=destroy`*(cache: LruCache) =
  ## Destructor: automatically frees the underlying C LRU cache.
  if cache.handle != nil:
    provenLruFree(cache.handle)

proc `=copy`*(dest: var LruCache, src: LruCache) {.error:
  "LruCache cannot be copied; it owns an opaque C resource".}

proc `=sink`*(dest: var LruCache, src: LruCache) =
  ## Move semantics for LruCache.
  `=destroy`(dest)
  dest.handle = src.handle

proc newLruCache*(capacity: int): LruCache =
  ## Create a new LRU cache with the given capacity.
  ## Maximum capacity is 100,000 as enforced by libproven.
  let handle = provenLruCreate(csize_t(capacity))
  if handle == nil:
    raise newException(LruCacheError,
      "Failed to create LRU cache (allocation or invalid capacity)")
  LruCache(handle: handle)

proc get*(cache: LruCache, key: uint64): Option[int64] =
  ## Get a value by key, promoting it to most recently used.
  ## Returns None if key is not found.
  if cache.handle == nil:
    return none(int64)
  let res = provenLruGet(cache.handle, key)
  if res.status == PROVEN_OK:
    some(res.value)
  else:
    none(int64)

proc put*(cache: LruCache, key: uint64, value: int64): bool =
  ## Insert or update a key-value pair.  If the cache is full, the least
  ## recently used entry is evicted.  Returns true on success.
  if cache.handle == nil:
    return false
  let status = provenLruPut(cache.handle, key, value)
  status == PROVEN_OK

proc contains*(cache: LruCache, key: uint64): bool =
  ## Check if a key exists in the cache.
  ## Note: this does NOT promote the entry (it calls get which does promote,
  ## but there is no peek in the C API).
  if cache.handle == nil:
    return false
  let res = provenLruGet(cache.handle, key)
  res.status == PROVEN_OK

proc len*(cache: LruCache): int =
  ## Get current number of entries in the cache.
  if cache.handle != nil:
    int(cache.handle[].count)
  else:
    0

proc capacity*(cache: LruCache): int =
  ## Get the cache capacity.
  if cache.handle != nil:
    int(cache.handle[].capacity)
  else:
    0

proc isEmpty*(cache: LruCache): bool =
  ## Check if cache is empty.
  cache.len == 0

proc isFull*(cache: LruCache): bool =
  ## Check if cache is at capacity.
  if cache.handle != nil:
    cache.handle[].count >= cache.handle[].capacity
  else:
    true

proc isValid*(cache: LruCache): bool =
  ## Check if the cache handle is valid (non-nil).
  cache.handle != nil
